src/bpftrace.cpp

```diff
@@ -1308,10 +1331,75 @@ int BPFtrace::setup_perf_events()
   return 0;
 }
 
-void BPFtrace::poll_perf_events(bool drain)
+int BPFtrace::setup_ringbuf()
+{
+  ringbuf_ = static_cast<struct ring_buffer *>(bpf_new_ringbuf(
+      maps[MapManager::Type::Ringbuf].value()->mapfd_, ringbuf_printer, this));
+  // init ringbuf loss counter
+  if (bpf_update_elem(
+          maps[MapManager::Type::RingbufLossCounter].value()->mapfd_,
+          &rb_loss_cnt_key_,
+          &rb_loss_cnt_val_,
+          0))
+  {
+    LOG(ERROR) << "fail to init ringbuf loss counter";
+    return -1;
+  }
+  return 0;
+}
+
+void BPFtrace::teardown_output()
 {
-  const int timeout_ms = 100;
+  if (feature_->has_map_ringbuf())
+  {
+    bpf_free_ringbuf(ringbuf_);
+    // print ringbuf event losses
+    uint64_t value = 0;
+    if (bpf_lookup_elem(
+            maps[MapManager::Type::RingbufLossCounter].value()->mapfd_,
+            &rb_loss_cnt_key_,
+            &value))
+    {
+      LOG(ERROR) << "fail to get ringbuf loss counter";
+    }
+    if (value)
+    {
+      out_->lost_events(value);
+    }
+  }
 
+  if (!feature_->has_map_ringbuf() || resources.needs_perf_event_map)
+    // Calls perf_reader_free() on all open perf buffers.
+    open_perf_buffers_.clear();
+}
+
+void BPFtrace::poll_output(bool drain)
+{
+  if (feature_->has_map_ringbuf())
+  {
+    if (resources.needs_perf_event_map)
+    {
+      // poll events from ringbuf and perf_event at the same time.
+      std::thread t;
+      t = std::thread(&bpftrace::BPFtrace::poll_ringbuf, this, drain);
+      poll_perf_events(drain);
+      t.join();
+    }
```

Comment:
Rather than do the thread dance, please check if we can register the ringbuf fd with the epoll in `poll_perf_events()`. Then we can just rename `poll_perf_events()` and reduce the duplicated code and remove multi-threading. I'd rather not have to think about thread safety if possible.